### PR TITLE
[RESTEASY-1138] keep a reference to the ClientResponse object as long…

### PR DESCRIPTION
**Pull request by asoldano**
… as there's a live inputstream derived from it

### CODE REVIEW

File: resteasy-client/src/main/java/org/jboss/resteasy/client/jaxrs/internal/ClientResponse.java [hunk 1]
```diff
@@ -129,7 +130,24 @@ protected InputStream getEntityStream()
...
+      
+      protected InputStreamWrapper(InputStream in, ClientResponse response) {
+         super(in);
+         this.response = response;
+      }
+      
+      @Override
+      public void close() throws IOException {
+         super.close();
+         this.response.close();
```
Comment (asoldano) [inline resteasy-client/src/main/java/org/jboss/resteasy/client/jaxrs/internal/ClientResponse.java]: I could have simply set this.response = null here, but that would have been possibly "optimized" (removed) by an aggressive JIT compiler figuring out this.response wouldn't be used. So I'm actually explicitly closing the response if the inputstream is explicitly closed.

Comment (jimma) [inline resteasy-client/src/main/java/org/jboss/resteasy/client/jaxrs/internal/ClientResponse.java]: This PR looks good to me. One of the minor things: is the super.close() is enough here ?  response.close() will close the underlying inputstream and other stuff too.

File: testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/response/ResponseStreamPrematurelyClosedTest.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+   }
+
+   @Test
+   public void testStream() throws Exception
+   {
+      Builder builder = client.target(generateURL("/test/document/abc/content")).request();
+      ByteArrayOutputStream baos = new ByteArrayOutputStream();
+      //builder.get().readEntity explicitly on the same line below and not saved in any temp variable
+      //to let the JVM try finalizing the ClientResponse object
+      IOUtils.copy(builder.get().readEntity(InputStream.class), baos);
```
Comment (jimma) [inline testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/response/ResponseStreamPrematurelyClosedTest.java]: Another minor thing :  we can call a system.gc() (although it only suggest jvm to do gc) and wait for  gc notification by calling javax.management api. After gc happens , we do IOUtils.copy and check the target stream. Is it better ?

File: resteasy-client/src/main/java/org/jboss/resteasy/client/jaxrs/internal/ClientResponse.java (hunk 1 above)
Comment (asoldano) [inline resteasy-client/src/main/java/org/jboss/resteasy/client/jaxrs/internal/ClientResponse.java]: For the records, I've been discussing this a bit with Jim and we eventually agreed that it should be fine to leave the fix as is.

Comment (asoldano) [inline resteasy-client/src/main/java/org/jboss/resteasy/client/jaxrs/internal/ClientResponse.java]: For the records, I've been discussing this a bit with Jim and we eventually agreed that it should be fine to leave the fix as is.

File: testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/response/ResponseStreamPrematurelyClosedTest.java (hunk 2 above)
Comment (asoldano) [inline testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/response/ResponseStreamPrematurelyClosedTest.java]: Discussed this too, Jim might propose an additional test later.

**Comment by coxon** [comment]
hello, i update resteasy to 3.0.23.Final, but i still get this exception. if you need, i can post exception.

**Comment by asoldano** [comment]
@coxon , sure, please share as much details as you can.
Thanks

**Comment by coxon** [comment]
@asoldano  ok, i try to describe what i faced clearly.  
i used openstack4j to download objects, and openstack4j send http-post via reateasy. ApacheHttpClient4Executor-> ClientResponse execute(ClientRequest request) and i get Stream closed exception throws out.
it's happend occasionally,  so next time when exception  throws , i will post it.

**Pull request merged**